base/essentials.jl

```diff
@@ -30,15 +30,24 @@ end
     @nospecialize
 
 Applied to a function argument name, hints to the compiler that the method
-should not be specialized for different types of that argument.
+should not be specialized for different types of that argument,
+but instead to use precisely the declared type for each argument.
 This is only a hint for avoiding excess code generation.
-Can be applied to an argument within a formal argument list, or in the
-function body.
-When applied to an argument, the macro must wrap the entire argument
-expression.
+Can be applied to an argument within a formal argument list,
+or in the function body.
+When applied to an argument, the macro must wrap the entire argument expression.
 When used in a function body, the macro must occur in statement position and
 before any code.
 
+When used without arguments, it applies to all arguments of the parent scope.
+In local scope, this means all arguments of the containing function.
+In global (top-level) scope, this means all methods subsequently defined in the current module.
```

Comment:
It seems more natural to me to use this like
```
@nospecialize foo(...) = ... # applies to all arguments
@nospecialize begin ... end # applies to all methods defined in this block
@nospecialize module Foo ... end # applies to all methods in module Foo
```

Reply:
Sure, all of those can be added later to the macro. They'd just be syntax transforms to one of these forms, since this form – as an imperative side-effect – represents where that information needs to be present in the processing order (since method declaration is itself also imperative).

Reply:
FWIW, `@nospecialize false` is double negation.

Reply:
It does seem clearer for the low-level control to be `@specialize! true|false` defaulting to `true` to avoid the double negative. Then `@nospecialize` block forms like @stevengj outlined can be built on top of that.

Reply:
Yes, for the module-level form `@specialize` is better, but I'd keep `@nospecialize` as well since it's much more convenient when annotating a single argument. Then we could potentially use those instead of the true/false argument.

Reply:
I can’t tell if you’re trolling